lib/ansible/modules/cloud/amazon/aws_api_gateway.py

```diff
@@ -0,0 +1,337 @@
+            stage=dict(type='str', default=None),
+            deploy_desc=dict(type='str', default="Automatic deployment by Ansible."),
+        )
+    )
+
+    mutually_exclusive = [['swagger_file', 'swagger_dict', 'swagger_text']]  # noqa: F841
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False,
+                           mutually_exclusive=mutually_exclusive)
+
+    api_id = module.params.get('api_id')
+    state = module.params.get('state')   # noqa: F841
+    swagger_file = module.params.get('swagger_file')
+    swagger_dict = module.params.get('swagger_dict')
+    swagger_text = module.params.get('swagger_text')
+    stage = module.params.get('stage')
+    deploy_desc = module.params.get('deploy_desc')
+
+#    check_mode = module.check_mode
+    changed = False
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='Python module "boto3" is missing, please install boto3')
+
+    if not HAS_BOTOCORE:
+        module.fail_json(msg='Python module "botocore" is missing, please install it')
+
+    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+    try:
+        client = boto3_conn(module, conn_type='client', resource='apigateway',
+                            region=region, endpoint=ec2_url, **aws_connect_kwargs)
+    except botocore.exceptions.NoRegionError:
+        module.fail_json(msg="Region must be specified as a parameter, in "
+                         "AWS_DEFAULT_REGION environment variable or in boto configuration file")
+    except (botocore.exceptions.ValidationError, botocore.exceptions.ClientError) as e:
+        fail_json_aws(module, e, msg="connecting to AWS")
+
+    changed = True   # for now it will stay that way until we can sometimes avoid change
+
+    conf_res = None
+    dep_res = None
+    del_res = None
+
+    if state == "present":
+        if api_id is None:
+            api_id = create_empty_api(module, client)
+        api_data = get_api_definitions(module, swagger_file=swagger_file,
+                                       swagger_dict=swagger_dict, swagger_text=swagger_text)
+        conf_res, dep_res = ensure_api_in_correct_state(module, client, api_id=api_id,
+                                                        api_data=api_data, stage=stage,
+                                                        deploy_desc=deploy_desc)
+    if state == "absent":
+        del_res = delete_rest_api(module, client, api_id)
+
+    exit_args = {"changed": changed, "api_id": api_id}
+
+    if conf_res is not None:
+        exit_args['configure_response'] = camel_dict_to_snake_dict(conf_res)
+    if dep_res is not None:
+        exit_args['deploy_response'] = camel_dict_to_snake_dict(dep_res)
+    if del_res is not None:
+        exit_args['delete_response'] = camel_dict_to_snake_dict(del_res)
+
+    module.exit_json(**exit_args)
+
+
+def get_api_definitions(module, swagger_file=None, swagger_dict=None, swagger_text=None):
```

Comment:
I know we tend to ask for main to be the last function. I don't see that as a blocker (but not sure).